packages/react-query/src/utils/structuralSharing.ts

```diff
@@ -0,0 +1,148 @@
+/**
+ * Inspired by @tanstack/query-core utils.ts
+ * Modified to support `Date`, `Map`, `Set` and `NaN` comparisons
+ *
+ * This function returns `a` if `b` is deeply equal.
+ * If not, it will replace any deeply equal children of `b` with those of `a`.
+ *
+ * @example
+ * const structuralSharing = createStructuralSharingFunction((prev, next) => {
+ *   if (prev instanceof Temporal.PlainDate && next instanceof Temporal.PlainDate) {
+ *     return prev.equals(next);
+ *   }
+ *   return false
+ * });
+ */
+export function createStructuralSharingFunction(
+  customEqualityFunction: (a: unknown, b: unknown) => boolean,
+) {
+  const equalCheck = (a: unknown, b: unknown) => {
+    return customEqualityFunction(a, b) || defaultIsEqual(a, b);
+  };
+  const structuralSharingFunction = (prev: any, next: any): any => {
+    if (equalCheck(prev, next)) {
+      return prev;
+    }
+
+    const array = isPlainArray(prev) && isPlainArray(next);
+    const plainObject = isPlainObject(prev) && isPlainObject(next);
+    const map = prev instanceof Map && next instanceof Map;
+    const set = prev instanceof Set && next instanceof Set;
+
+    if (array || plainObject || map) {
```

Comment:
The code within here is.... horribly hard to follow

should we split it up into less DRY operations instead?